### PR TITLE
Draft: Fix for GHI-4644: mesh optimization breaking skin influences

### PR DESCRIPTION
**Pull request by amzn-tommy**
There are two issues that cropped up from treating skin influences separately from the rest of the vertex data

One is that the vertices are 'de-duplicated' if they have the same vertex data but unique skin influences, since the skin influences aren't part of the layer data that is used during de-duplication. I created an example of that
Before:

VerticesWithDifferentJointIdsShouldNotWeld_Before

After:

VerticesWithDifferentJointIdsShouldNotWeld_After

The other issue was that when vertices should legitimately be de-duplicated, the skinning info was stacking multiple copies of the same influence onto a single vertex. So if the vertex had two influences
{0, .75}
{1, .25}
The optimizer would see
{{0, .75}, {1, .25}, {0, .75}, {1, .25}, {0, .75}, {1, .25}, {0, .75}, {1, .25}}
And it would get optimized to
{0, 1.0}
because it sorts by weight, putting all the {0, .75}'s in front, then cuts the lowest weight links (which is all the {1, .25} links), then re-normalizes a vector of four {0, .75} influences to a single {0, 1.0} influence.
Here is an example of that:
Before:

WeldedVerticesShouldMaintainOriginalSkinInfluences_Before

After:
WeldedVerticesShouldMaintainOriginalSkinInfluences_After

And the original extruding vertex on Valena is resolved
Before:

Valena_Before

After:

Valena_After

### CODE REVIEW

File: Gems/SceneProcessing/Code/Source/Generation/Components/MeshOptimizer/MeshOptimizerComponent.cpp [hunk 1]
```diff
@@ -639,6 +696,19 @@ namespace AZ::SceneGenerationComponents
                 {
                     optimizedVertexColorNode->AppendColor(vertexColorLayer->GetVertexValue(vertexLookup.mOrgVtx, vertexLookup.mDuplicateNr));
                 }
+
+                if (optimizedSkinWeights)
+                {
+                    AZStd::vector<AZ::MeshBuilder::MeshBuilderSkinningInfo::Influence> influences =
+                        ExtractSkinningInfo(skinningInfluencesLayers, vertexLookup, maxWeightsPerVertex, weightThreshold);
+
+                    for (auto influence : influences)
```
Comment (burelc-amzn) [inline Gems/SceneProcessing/Code/Source/Generation/Components/MeshOptimizer/MeshOptimizerComponent.cpp:705]: Nit:
```suggestion
                    for (const auto& influence : influences)
```

File: Gems/SceneProcessing/Code/Tests/MeshBuilder/MeshOptimizerComponentTests.cpp [hunk 2]
```diff
@@ -94,7 +110,28 @@ namespace SceneProcessing
...
+            auto skinWeights = AZStd::make_unique<AZ::SceneData::GraphData::SkinWeightData>();
+
+            skinWeights->ResizeContainerSpace(sourceLinks.size());
+
+            for (size_t vertexIndex = 0; vertexIndex < sourceLinks.size(); ++vertexIndex)
+            {
+                for (const auto& link : sourceLinks[vertexIndex])
+                {
+                    // Make sure the bone is added to the skin weights
+                    skinWeights->GetBoneId(AZStd::string::format("%d", link.boneId));
```
Comment (burelc-amzn) [inline Gems/SceneProcessing/Code/Tests/MeshBuilder/MeshOptimizerComponentTests.cpp:125]: Nit:
```suggestion
                    skinWeights->GetBoneId(AZStd::to_string(link.boneId));
```

**Review by burelc-amzn — APPROVED** [review]
Looks good!

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Model/ModelAssetBuilderComponent.cpp [hunk 3]
```diff
@@ -948,19 +948,17 @@ namespace AZ
...
             AZStd::vector<uint16_t>& skinJointIndices = productMesh.m_skinJointIndices;
             AZStd::vector<float>& skinWeights = productMesh.m_skinWeights;
-            const auto& sourceMeshData = sourceMesh.m_meshData;
 
             size_t numInfluencesAdded = 0;
             for (const auto& skinData : sourceMesh.m_skinData)
             {
-                const AZ::u32 controlPointIndex = sourceMeshData->GetControlPointIndex(static_cast<int>(vertexIndex));
-                const size_t numSkinInfluences = skinData->GetLinkCount(controlPointIndex);
+                const size_t numSkinInfluences = skinData->GetLinkCount(vertexIndex);
```
[the remark this answers is not in the document]
Comment (amzn-tommy) [reply Gems/Atom/RPI/Code/Source/RPI.Builders/Model/ModelAssetBuilderComponent.cpp:955]: This is a great question, and I will test it out. But my intuition is yes, it is still correct without mesh optimization. Assimp is generating a 1-1 vertex index to control point index mapping, so functionally it's actually no different than what was there before. And fundamentally, skin influences are per-vertex data, not per-control point data. Only positions are per-control point. So this should have been using vertexIndex all along.

**Review by amzn-jillich — APPROVED** [review]

**Comment by amzn-tommy** [comment]
Submitted in another PR, from a new branch to fix the DCO failures https://github.com/o3de/o3de/pull/6336

**Pull request closed without merging**